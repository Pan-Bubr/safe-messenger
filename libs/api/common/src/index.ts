export * from './lib/api-common.module';

export * from './lib/entities/user-entity';
export * from './lib/entities/connection-entity';
export * from './lib/entities/message-entity';

export * from './lib/services/user.service';
export * from './lib/services/connection.service';
export * from './lib/services/crypto.service';
export * from './lib/services/message.service';
