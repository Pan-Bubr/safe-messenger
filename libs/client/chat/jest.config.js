module.exports = {
  name: 'client-chat',
  preset: '../../../jest.config.js',
  coverageDirectory: '../../../coverage/libs/client/chat',
  snapshotSerializers: [
    'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer.js',
    'jest-preset-angular/build/AngularSnapshotSerializer.js',
    'jest-preset-angular/build/HTMLCommentSerializer.js'
  ],
  coverageThreshold: {
    global: {
      functions: 60,
      branches: 70,
      lines: 70,
      statements: 70
    }
  },
};
