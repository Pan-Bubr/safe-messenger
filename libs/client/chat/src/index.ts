export * from './lib/client-chat.module';
