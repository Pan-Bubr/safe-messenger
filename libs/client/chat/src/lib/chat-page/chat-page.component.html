<div class="p-5">
  <div class="row">
    <button class="btn btn-sm btn-dark" (click)="returnToHomepage()">
      Return to Home Page
    </button>
  </div>

  <h3 class="row mt-3">Chat with {{ recipient$ | async }}</h3>
  <p class="row text-muted">
    <small>
      If you suspect your keys may have been leaked, create a new identity.
      <button class="btn btn-sm btn-danger" (click)="resetConnection()">
        Create new identity
      </button>
    </small>
  </p>
  <div class="row"></div>

  <div
    *ngFor="let message of historyMessages$ | async; let i = index"
    class="message"
    [ngClass]="{ 'text-right': (recipient$ | async) === message.senderEmail }"
  >
    <div class="">
      <small>
        {{ message.sentAt | date: 'shortTime' }}
      </small>
      <div>
        <div class="d-inline-block border rounded p-2">
          {{ message.content }}
        </div>
      </div>
    </div>
  </div>

  <div
    *ngFor="let message of newMessages"
    class="message"
    [ngClass]="{ 'text-right': (recipient$ | async) === message.senderEmail }"
  >
    <div class="">
      <small>
        {{ message.sentAt | date: 'shortTime' }}
      </small>
      <div *ngIf="message.content">
        <div class="d-inline-block border rounded p-2">
          {{ message.content }}
        </div>
      </div>
    </div>
  </div>

  <div class="row mt-4">
    <form [formGroup]="chatForm" (ngSubmit)="sendMessage(chatForm.value)">
      <div class="input-group mb-3">
        <div class="input-group-prepend">
          <span class="input-group-text">Send Message:</span>
        </div>
        <input
          name="message"
          formControlName="message"
          type="text"
          class="form-control"
          placeholder="Enter message here..."
          required
        />

        <div class="input-group-append">
          <button
            class="btn btn-primary"
            type="submit"
            [disabled]="!chatForm.valid"
          >
            Send
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
