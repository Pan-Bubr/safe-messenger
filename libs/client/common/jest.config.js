module.exports = {
  name: 'client-common',
  preset: '../../../jest.config.js',
  coverageDirectory: '../../../coverage/libs/client/common',
  snapshotSerializers: [
    'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer.js',
    'jest-preset-angular/build/AngularSnapshotSerializer.js',
    'jest-preset-angular/build/HTMLCommentSerializer.js'
  ],
  coverageThreshold: {
    global: {
      functions: 0,
      branches: 0,
      lines: 0,
      statements: 0
    }
  }
};
