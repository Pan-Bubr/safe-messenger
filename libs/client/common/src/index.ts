export * from './lib/client-common.module';
export * from './lib/auth.interceptor';

export * from './lib/services/contacts.service';
