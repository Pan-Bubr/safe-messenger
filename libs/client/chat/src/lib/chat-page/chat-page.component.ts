import { Component, OnDestroy, OnInit } from '@angular/core';
import { FormBuilder, FormControl, FormGroup } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { ErrorMessage, RawMessage } from '@safe-messenger/api-interfaces';
import { ContactsService } from '@safe-messenger/client/common';

import { Observable } from 'rxjs';
import { pluck, switchMap } from 'rxjs/operators';

import { ChatService } from '../services/chat.service';

@Component({
  selector: 'safe-messenger-chat-page',
  templateUrl: './chat-page.component.html',
  styleUrls: ['./chat-page.component.scss']
})
export class ChatPageComponent implements OnInit, OnDestroy {
  constructor(
    private readonly route: ActivatedRoute,
    private readonly router: Router,
    private readonly chatService: ChatService,
    private readonly contactsService: ContactsService
  ) {}

  public historyMessages$: Observable<RawMessage[]> = this.route.params.pipe(
    pluck('email'),
    switchMap(email => this.chatService.getHistory(email))
  );

  public newMessage$: Observable<RawMessage | ErrorMessage> = this.chatService
    .newMessage$;

  public recipient$: Observable<string> = this.route.params.pipe(
    pluck('email')
  );
  public page: number = 0;
  public chatForm: FormGroup = new FormGroup({
    message: new FormControl('')
  });

  public newMessages: (RawMessage | ErrorMessage)[] = [];
  private sessionId: string;

  public async ngOnInit(): Promise<void> {
    this.sessionId = this.route.snapshot.data['sessionId'];
    this.recipient$.subscribe(async email => {
      await this.chatService.join(email);
      this.chatService.getServerMessages(email);
    });
    this.newMessage$.subscribe(message => {
      this.newMessages = [...this.newMessages, message];
    });
  }

  public async ngOnDestroy(): Promise<void> {
    await this.chatService.leave();
  }

  public sendMessage({ message }: { message: string }): void {
    this.chatForm.reset();
    if (message.trim()) {
      this.recipient$.subscribe(async recipientEmail => {
        await this.chatService.sendMessage(
          {
            senderEmail: localStorage.getItem('email'),
            recipientEmail,
            sessionId: this.sessionId,
            content: message.trim()
          },
          this.sessionId
        );
      });
    }
  }

  public returnToHomepage(): Promise<boolean> {
    return this.router.navigate(['home']);
  }

  public resetConnection(): void {
    this.recipient$.subscribe(async recipientEmail => {
      this.contactsService.resetConnection(recipientEmail).subscribe(() => {
        this.router.navigateByUrl('/');
      });
    });
  }
}
