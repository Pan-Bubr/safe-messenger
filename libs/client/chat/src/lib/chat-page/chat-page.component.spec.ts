import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { ClientChatModule } from '@safe-messenger/client/chat';
import { ContactsService } from '@safe-messenger/client/common';

import { of } from 'rxjs';
import { Shallow } from 'shallow-render';

import { ChatService } from '../services/chat.service';
import { ChatPageComponent } from './chat-page.component';

describe('ChatPageComponent', () => {
  let shallow: Shallow<ChatPageComponent>;

  beforeEach(() => {
    shallow = new Shallow(ChatPageComponent, ClientChatModule)
      .import(HttpClientTestingModule, ReactiveFormsModule)
      .provideMock(
        {
          provide: Router,
          useValue: {
            navigateByUrl: jest.fn(),
            navigate: jest.fn()
          }
        },
        {
          provide: ActivatedRoute,
          useValue: {
            params: of({ email: 'mockEmail' }),
            snapshot: {
              data: {
                sessionId: '1'
              }
            }
          }
        },
        {
          provide: ChatService,
          useValue: {
            join: jest.fn(),
            leave: jest.fn(),
            getHistory: jest.fn(() => of([])),
            getServerMessages: jest.fn(),
            sendMessage: jest.fn(),
            newMessage$: of({
              content: 'test',
              sessionId: 1,
              recipientEmail: 'mockTarget',
              senderEmail: 'mockUser'
            })
          }
        },
        {
          provide: ContactsService,
          useValue: {
            resetConnection: jest.fn(() => of(''))
          }
        }
      );
  });

  it('should match snapshot', async () => {
    const { fixture } = await shallow.render();

    expect(fixture).toMatchSnapshot();
  });

  it('ngOnInit()', async () => {
    const { instance } = await shallow.render();

    await instance.ngOnInit();
    await instance.ngOnDestroy();
  });

  it('sendMessage() should send Message', async () => {
    const { instance, get } = await shallow.render();

    instance.sendMessage({ message: 'text' });

    expect(get(ChatService).sendMessage).toBeCalled();
  });

  it('sendMessage() should not send empty Message', async () => {
    const { instance, get } = await shallow.render();

    instance.sendMessage({ message: '   ' });

    expect(get(ChatService).sendMessage).not.toBeCalled();
  });

  it('returnToHomepage() should return to homepage', async () => {
    const { instance, get } = await shallow.render();

    await instance.returnToHomepage();

    expect(get(Router).navigate).toBeCalled();
  });

  it('returnToHomepage() should return to homepage', async () => {
    const { instance, get } = await shallow.render();

    instance.resetConnection();

    expect(get(Router).navigateByUrl).toBeCalled();
  });
});
