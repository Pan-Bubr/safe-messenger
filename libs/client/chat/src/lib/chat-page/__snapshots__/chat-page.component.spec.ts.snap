// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatPageComponent should match snapshot 1`] = `
<ng-component>
  <safe-messenger-chat-page>
    <div
      class="p-5"
    >
      <div
        class="row"
      >
        <button
          class="btn btn-sm btn-dark"
        >
           Return to Home Page 
        </button>
      </div>
      <h3
        class="row mt-3"
      >
        Chat with mockEmail
      </h3>
      <p
        class="row text-muted"
      >
        <small>
           If you suspect your keys may have been leaked, create a new identity. 
          <button
            class="btn btn-sm btn-danger"
          >
             Create new identity 
          </button>
        </small>
      </p>
      <div
        class="row"
      />
      
      
      <div
        class="message"
      >
        <div
          class=""
        >
          <small>
              
          </small>
          
          <div>
            <div
              class="d-inline-block border rounded p-2"
            >
               test 
            </div>
          </div>
        </div>
      </div>
      <div
        class="row mt-4"
      >
        <form
          class="ng-untouched ng-pristine ng-invalid"
          novalidate=""
        >
          <div
            class="input-group mb-3"
          >
            <div
              class="input-group-prepend"
            >
              <span
                class="input-group-text"
              >
                Send Message:
              </span>
            </div>
            <input
              class="form-control ng-untouched ng-pristine ng-invalid"
              formcontrolname="message"
              name="message"
              placeholder="Enter message here..."
              required=""
              type="text"
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-primary"
                disabled=""
                type="submit"
              >
                 Send 
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </safe-messenger-chat-page>
</ng-component>
`;
